src/fifelse.c

```diff
@@ -201,105 +201,106 @@ SEXP fifelseR(SEXP l, SEXP a, SEXP b, SEXP na) {
-      nonna = false;
-    } else {
-      if (tn != type0) {
-        error(_("Resulting value is of type %s but 'default' is of type %s. "
-                "Please make sure that both arguments have the same type."), type2char(type0), type2char(tn));
-      }
-      if (!R_compute_identical(PROTECT(getAttrib(value0,R_ClassSymbol)),  PROTECT(getAttrib(na,R_ClassSymbol)), 0)) {
-        error(_("Resulting value has different class than 'default'. "
-                "Please make sure that both arguments have the same class."));
-      }
-      UNPROTECT(2);
-      if (isFactor(value0)) {
-        if (!R_compute_identical(PROTECT(getAttrib(value0,R_LevelsSymbol)),  PROTECT(getAttrib(na,R_LevelsSymbol)), 0)) {
-          error(_("Resulting value and 'default' are both type factor but their levels are different."));
-        }
-        UNPROTECT(2);
-      }
-    }
-  }
-
+  int nprotect=0, l=0;
+  int64_t len0=0, len1=0, len2=0;
+  SEXP ans=R_NilValue, value0=R_NilValue, tracker=R_NilValue, whens=R_NilValue, thens=R_NilValue;
+  PROTECT_INDEX Iwhens, Ithens;
+  PROTECT_WITH_INDEX(whens, &Iwhens); nprotect++;
+  PROTECT_WITH_INDEX(thens, &Ithens); nprotect++;
+  SEXPTYPE type0=NILSXP;
+  // naout means if the output is scalar logic na
+  bool imask = true, naout = false, idefault = false;
+  int *restrict p = NULL;
   const int n = narg/2;
   for (int i=0; i<n; ++i) {
-    SEXP cons = PROTECT(i==0 ? cons0 : eval(SEXPPTR_RO(args)[2*i], rho)); // protect cons0 again for easy unprotect at the end of this loop
-    SEXP outs = PROTECT(i==0 ? value0 : eval(SEXPPTR_RO(args)[2*i+1], rho));
-    if (isS4(outs) && !INHERITS(outs, char_nanotime)) {
+    idefault = i == (n - 1); // mark if the current eval is the `default` on R side
+    REPROTECT(whens = eval(SEXPPTR_RO(args)[2*i], rho), Iwhens);
+    REPROTECT(thens = eval(SEXPPTR_RO(args)[2*i+1], rho), Ithens);
+    if (isS4(thens) && !INHERITS(thens, char_nanotime)) {
       error(_("S4 class objects (except nanotime) are not supported. Please see https://github.com/Rdatatable/data.table/issues/4131."));
     }
-    if (!isLogical(cons)) {
-      error(_("Argument #%d must be logical."), 2*i+1);
+    if (!isLogical(whens)) {
+      error(_("Argument #%d must be logical but was of type %s."), 2*i+1, type2char(TYPEOF(whens)));
     }
-    if (i>0) {
-      if (xlength(cons) != len0) {
-        error(_("Argument #%d has a different length than argument #1. "
-                "Please make sure all logical conditions have the same length."),
-                i*2+1);
+    const int *restrict pwhens = LOGICAL(whens);
+    if (i == 0) {
+      len0 = xlength(whens);
+      len2 = len0;
+      type0 = TYPEOF(thens);
+      value0 = thens;
+      ans = PROTECT(allocVector(type0, len0)); nprotect++;
+      copyMostAttrib(thens, ans);
+      tracker = PROTECT(allocVector(INTSXP, len0)); nprotect++;
+      p = INTEGER(tracker);     
+    } else {
+      imask = false;
+      l = 0;
```

Comment:
should be moved before `if (i==0) {} else {}` branching

Reply:
I think you're just referring to `l=0`; It works because `l=0` is also set to the initial value.

I agree it's clearer to just always initialize `l=0` since it's needed in each iteration.